resources/providers/azurelib/inventory/keyvault_provider.go

```diff
@@ -0,0 +1,190 @@
+				return nil, err
+			}
+			return readPager(ctx, client.NewListPager(resourceGroupName, vaultName, nil))
+		},
+		AssetKeyVaultSecrets: func(ctx context.Context, subscriptionID string, resourceGroupName string, vaultName string) ([]armkeyvault.SecretsClientListResponse, error) {
+			client, err := armkeyvault.NewSecretsClient(subscriptionID, credentials, nil)
+			if err != nil {
+				return nil, err
+			}
+			return readPager(ctx, client.NewListPager(resourceGroupName, vaultName, nil))
+		},
+	}
+}
+
+type KeyVaultProviderAPI interface {
+	ListKeyVaultKeys(ctx context.Context, vault AzureAsset) ([]AzureAsset, error)
+	ListKeyVaultSecrets(ctx context.Context, vault AzureAsset) ([]AzureAsset, error)
+}
+
+func NewKeyVaultProvider(log *logp.Logger, credentials azcore.TokenCredential) KeyVaultProviderAPI {
+	return &keyVaultProvider{
+		log:    log,
+		client: defaultAzureKeyVaultWrapper(credentials),
+	}
+}
+
+type keyVaultProvider struct {
+	log    *logp.Logger
+	client *azureKeyVaultWrapper
+}
+
+func (p *keyVaultProvider) ListKeyVaultKeys(ctx context.Context, vault AzureAsset) ([]AzureAsset, error) {
+	p.log.Info("Listing Azure Vault Keys")
+
+	responses, err := p.client.AssetKeyVaultKeys(ctx, vault.SubscriptionId, vault.ResourceGroup, vault.Name)
+	if err != nil {
+		return nil, fmt.Errorf("error while retrieving vault keys: %w", err)
+	}
+
+	return lo.FlatMap(responses, func(res armkeyvault.KeysClientListResponse, _ int) []AzureAsset {
+		return lo.FilterMap(res.Value, func(key *armkeyvault.Key, _ int) (AzureAsset, bool) {
+			return p.transformKey(key, vault)
+		})
+	}), nil
+}
+
+func (p *keyVaultProvider) transformKey(key *armkeyvault.Key, vault AzureAsset) (AzureAsset, bool) {
+	if key == nil {
+		return AzureAsset{}, false
+	}
+
+	properties := map[string]any{}
+
+	addKeyIfMapNotEmpty(properties, "tags", key.Tags)
+	addKeyIfNotEmpty(properties, "attributes", key.Properties.Attributes)
+	addKeyIfNotEmpty(properties, "curveName", key.Properties.CurveName)
+	addKeyIfSliceNotEmpty(properties, "keyOps", key.Properties.KeyOps)
+	addKeyIfNotEmpty(properties, "keySize", key.Properties.KeySize)
+	addKeyIfNotEmpty(properties, "keyUri", key.Properties.KeyURI)
+	addKeyIfNotEmpty(properties, "keyUriWithVersion", key.Properties.KeyURIWithVersion)
+	addKeyIfNotEmpty(properties, "kty", key.Properties.Kty)
+	addKeyIfNotEmpty(properties, "releasePolicy", key.Properties.ReleasePolicy)
+	addKeyIfNotEmpty(properties, "rotationPolicy", key.Properties.RotationPolicy)
+	if len(properties) == 0 {
+		properties = nil
+	}
+
+	return AzureAsset{
+		Id:             pointers.Deref(key.ID),
+		Name:           pointers.Deref(key.Name),
+		DisplayName:    "",
+		Location:       pointers.Deref(key.Location),
+		ResourceGroup:  vault.ResourceGroup,
+		SubscriptionId: vault.SubscriptionId,
+		TenantId:       vault.TenantId,
+		Type:           pointers.Deref(key.Type),
+		Properties:     properties,
+		Extension: map[string]any{
+			ExtensionVaultID: vault.Id,
```

Comment:
I find it interesting that this Asset, which is an extension already, has an extension. Why is that?

Reply:
And I could not find a usage of the extension key `vaultId`

Reply:
Good catch; this was left there by mistake (the first approach was to group all resources together and then, using this extension field, merge with each key vault, like in the storage account). 
But then changed to provider call per asset approach so this is unnecessary. 
I will remove it.